src/tss2-fapi/ifapi_ima_eventlog.c

```diff
@@ -0,0 +1,853 @@
+    size_t size;
+    UINT32 size_from_buffer;
+    (void)digest;
+    (void)jso;
+
+    /* Get size from buffer without 0 Terminator. */
+    size_from_buffer = get_size_from_buffer(buffer, offset) - 1;
+    size = strlen((const char *)&buffer[*offset]);
+    if (size != size_from_buffer) {
+        return_error2(TSS2_FAPI_RC_BAD_VALUE,
+                      "Invalid digest size, string length: %zu size from buffer: %"
+                      PRIu32, size, size_from_buffer);
+    }
+    LOG_TRACE("IMA name: %s", (const char *)&buffer[*offset]);
+    template->name = (char *)&buffer[*offset];
+    *offset += size + 1; /**< with 0 terminator */
+    return TSS2_RC_SUCCESS;
+}
+
+/** Callback to get null terminated name (n).
+ */
+static TSS2_RC eventname_cb(UINT8 *digest, UINT8 * buffer, size_t *offset, json_object *jso,
+                            IFAPI_IMA_TEMPLATE *template) {
+    size_t size;
+    (void)digest;
+    (void)jso;
+
+    size = strlen((const char *)&buffer[*offset]); // TODO check
+    if (size > TCG_EVENT_NAME_LEN_MAX + 1) {
+        return_error(TSS2_FAPI_RC_BAD_VALUE, "Too long event name.");
+    }
+    LOG_TRACE("IMA name: %s", (const char *)&buffer[*offset]);
+    template->name =  (char *)&buffer[*offset];
+    *offset += size + 1; /**< with 0 terminator */
+    return TSS2_RC_SUCCESS;
+}
+
+/** Callback to initialize the json event list.
+ */
+TSS2_RC
+init_event_list_json_cb(json_object **jso) {
+    return_if_null(jso, "Bad reference.", TSS2_FAPI_RC_BAD_VALUE);
+    if (!*jso) {
+        *jso = json_object_new_array();
+        return_if_null(*jso, "Out of memory", TSS2_FAPI_RC_MEMORY);
+    }
+    return TSS2_RC_SUCCESS;
+}
+
+/** Callback to convert header of IMA template to JSON.
+ */
+TSS2_RC
+event_header_json_cb(
+    size_t recnum,
+    UINT32 pcr,
+    const char* ima_type,
+    UINT8 *digest,
+    size_t digest_size,
+    json_object *jso_list,
+    json_object **jso)
+{
+    TSS2_RC r;
+    json_object *jso_digest, *jso_digest_type, *jso_ary, *jso_content;
+    char *hash_name;
+
+    *jso = json_object_new_object();
+    return_if_null(*jso, "Out of memory.", TSS2_FAPI_RC_MEMORY);
+
+    r = add_number_to_json(recnum, *jso, "recnum");
+    return_if_error(r, "Add number to json object.");
+
+    r = add_number_to_json(pcr, *jso, "pcr");
+    return_if_error(r, "Add number to json object.");
+
+    jso_digest = json_object_new_object();
+    return_if_null(*jso, "Out of memory.", TSS2_FAPI_RC_MEMORY);
+
+    jso_digest_type = NULL;
+    switch (digest_size) {
+        case TPM2_SHA1_DIGEST_SIZE:
+            hash_name = "sha1";
+            break;
+        case TPM2_SHA256_DIGEST_SIZE:
+            hash_name = "sha256";
+            break;
+        case TPM2_SHA384_DIGEST_SIZE:
+            hash_name = "sha384";
+            break;
+        case TPM2_SHA512_DIGEST_SIZE:
+            hash_name = "sha512";
+            break;
+        default:
```

Comment:
Won't all these assumptions break on SM3? As SM3 IIRC is 256 bits?

Reply:
sm3 is not possible for the template-hash it should be sha1. The function is prepared for cases where e.g. sha256 will be allowed.